pkg/scheduler/partition.go

```diff
@@ -0,0 +1,1101 @@
+	"github.com/apache/incubator-yunikorn-core/pkg/scheduler/policies"
+	"github.com/apache/incubator-yunikorn-core/pkg/webservice/dao"
+	"github.com/apache/incubator-yunikorn-scheduler-interface/lib/go/si"
+)
+
+type PartitionContext struct {
+	RmID string // the RM the partition belongs to
+	Name string // name of the partition (logging mainly)
+
+	// Private fields need protection
+	root                   *objects.Queue                  // start of the queue hierarchy
+	applications           map[string]*objects.Application // applications assigned to this partition
+	reservedApps           map[string]int                  // applications reserved within this partition, with reservation count
+	nodes                  map[string]*objects.Node        // nodes assigned to this partition
+	allocations            map[string]*objects.Allocation  // allocations
+	placementManager       *placement.AppPlacementManager  // placement manager for this partition
+	partitionManager       *partitionManager               // manager for this partition
+	stateMachine           *fsm.FSM                        // the state of the partition for scheduling
+	stateTime              time.Time                       // last time the state was updated (needed for cleanup)
+	isPreemptable          bool                            // can allocations be preempted
+	rules                  *[]configs.PlacementRule        // placement rules to be loaded by the scheduler
+	userGroupCache         *security.UserGroupCache        // user cache per partition
+	totalPartitionResource *resources.Resource             // Total node resources
+	nodeSortingPolicy      *policies.NodeSortingPolicy     // Global Node Sorting Policies
+
+	sync.RWMutex
+}
+
+func newPartitionContext(conf configs.PartitionConfig, rmID string, cc *ClusterContext) (*PartitionContext, error) {
+	if conf.Name == "" || rmID == "" {
+		log.Logger().Info("partition cannot be created",
+			zap.String("partition name", conf.Name),
+			zap.String("rmID", rmID),
+			zap.Any("cluster context", cc))
+		return nil, fmt.Errorf("partition cannot be created without name or RM, one is not set")
+	}
+	pc := &PartitionContext{
+		Name:         conf.Name,
+		RmID:         rmID,
+		stateMachine: objects.NewObjectState(),
+		stateTime:    time.Now(),
+		applications: make(map[string]*objects.Application),
+		reservedApps: make(map[string]int),
+		nodes:        make(map[string]*objects.Node),
+		allocations:  make(map[string]*objects.Allocation),
+	}
+	pc.partitionManager = &partitionManager{
+		pc: pc,
+		cc: cc,
+	}
+	if err := pc.initialPartitionFromConfig(conf); err != nil {
+		return nil, err
+	}
+	return pc, nil
+}
+
+// Initialise the partition
+func (pc *PartitionContext) initialPartitionFromConfig(conf configs.PartitionConfig) error {
+	if len(conf.Queues) == 0 || conf.Queues[0].Name != configs.RootQueue {
+		return fmt.Errorf("partition cannot be created without root queue")
+	}
```

Comment:
The root queue is inserted when the config is loaded. It will do a check of the config and based on that it will trigger an insert see `configs.checkQueuesStructure()` That has always been there. This is an extra safeguard to make sure the config is OK.